src/main/scala/chiseltest/formal/backends/Trace.scala

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiseltest.formal.backends
+
+import chiseltest.simulator.{Compiler, TreadleBackendAnnotation}
+import firrtl._
+
+/** Represents the inputs and starting state needed to re-create an execution trace. */
+private case class Trace(
+  inputs:  Seq[Seq[(String, BigInt)]],
+  regInit: Seq[(String, BigInt)],
+  memInit: Seq[(String, Seq[BigInt])])
+
+private object Trace {
+  def replayOnTreadle(trace: Trace, circuit: ir.Circuit, annos: AnnotationSeq): Unit = {
+    val dut = TreadleBackendAnnotation.getSimulator.createContext(CircuitState(circuit, annos))
+
+    // initialize
+    trace.regInit.foreach { case (name, value) => dut.poke(name, value) }
```

Comment:
So you have implemented the internal signal poke?

Reply:
This only works for treadle. Poking registers in Verilator is very difficult as it turns out and will probably never be implemented.

Reply:
I think here exist some problem:
I only see Data -> Name map expose via `DataMirror` rather than a FIRRTL pass, which make `reg` you poke not found in the map.

Reply:
I think this can happen via:
1. add a pass to extract `Data -> Verilog Name` via RenameMap, by default it can be port only, and let user add expose annotations.
2. generate a Verilator [config file](https://verilator.org/guide/latest/exe_verilator.html#cmdoption-arg-public) to expose those signals.
3. improve JNA/JNI interface to this.

Reply:
You could always try to prototype it, but from my experiments and looking at the Verilator generated code, the problem is that on a clock edge, Verilator not only updates the register values, it also calculated the next register values. Thus if you change a register after the clock edge, it won't affect the next value that was already calculated and thus your change is going to be overwritten at the next clock edge (unless the register isn't updated by the hardware).

Reply:
That is what I discovered when trying to use registers for coverage counter: https://github.com/ucb-bar/chisel-testers2/pull/353

Reply:
maybe [cocotb](https://github.com/cocotb/cocotb/blob/fba7b854b00375500552ff8804757f468152b111/cocotb/handle.py#L694-L698)  seems to be better to refer to?
It seems to use `Force` and `Release` to achieve this.